src/Containers/Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -447,32 +581,91 @@ private async Task<UriBuilder> StartUploadSessionAsync(string repository, string
-        return GetNextLocation(pushResponse);
+        var chunkSize = ParseRangeAmount(pushResponse) ?? ParseOCIChunkMinSizeAmount(pushResponse);
+        return new(chunkSize, GetNextLocation(pushResponse), IsAmazonECRRegistry);
     }
 
-    private Task<UriBuilder> UploadBlobContentsAsync(string repository, string digest, Stream contents, HttpClient client, UriBuilder uploadUri, CancellationToken cancellationToken)
+    /// <summary>
+    /// Computes the effective chunk size to use for the upload
+    /// </summary>
+    /// <remarks>
+    /// The chunk size is determined by the following rules:
+    /// We compare the registry's expressed chunk size with any user-provided chunk size (via env var).
+    /// If both are set, we use the smaller of the two.
+    /// If only one is set, we use that one.
+    /// If either is zero, we use the other.
+    /// If neither is set, we use the default chunk size.
+    /// Finally, AWS ECR has a min size that we use to override the above if it would be below that min size.
+    /// </remarks>
+    private static int EffectiveChunkSize(int? registryChunkSize, bool isAWS)
     {
-        cancellationToken.ThrowIfCancellationRequested();
-        if (SupportsChunkedUpload)
+
+        var result =
+            (registryChunkSize, s_chunkedUploadSizeBytes) switch
+            {
+                (0, int u) => u,
+                (int r, 0) => r,
+                (int r, int u) => Math.Min(r, u),
+                (int r, null) => r,
+                (null, int u) => u,
+                (null, null) => s_defaultChunkSizeBytes
+            };
+
+        if (isAWS)
         {
-            return UploadBlobChunkedAsync(repository, digest, contents, client, uploadUri, cancellationToken);
+            // AWS ECR requires a min chunk size of 5MB for all chunks except the last.
+            return Math.Max(result, s_fiveMegs);
         }
         else
         {
-            return UploadBlobWholeAsync(repository, digest, contents, client, uploadUri, cancellationToken);
+            return result;
         }
     }
 
-    private static async Task FinishUploadSessionAsync(string digest, HttpClient client, UriBuilder uploadUri, CancellationToken cancellationToken)
+    /// <summary>
+    /// Uploads the contents of the stream, either in an atomic PUT or in chunks.
+    /// </summary>
+    /// <remarks>
+    /// Generally, we try to upload the blob in an atomic PUT. If the registry does not support this, we fall back to chunked uploads.
+    /// If the atomic PUT fails, we fall back to chunked uploads.
+    /// If the registry provides a chunk max and that max is less than the content length, then we use chunked uploads.
+    /// </remarks>
+    private async Task<UploadFinalizeInformation> UploadBlobContentsAsync(string repository, string digest, Stream contents, HttpClient client, UploadInformation uploadInfo, Action<LogMessage> logProgressMessage, CancellationToken cancellationToken)
```

Comment:
Now that we've started an upload, we actually perform the upload. This method got _a lot_ smarter. It now tries to do an atomic upload if available - and it seems to be pretty widely available. If the registry actually defined a chunk size, and that chunk size is > than the layer to upload, we feel pretty confident that we can support an atomic upload.

However we also need to recover from faults and switch to a chunked upload for resiliency.